### ISSUE TITLE
docker-runc does not terminate and leave docker-shim hanging in 17.12

### ISSUE THREAD

**Issue by phsiao** [moby/moby#36010]
**Description**

This seems to be new to 17.12.

docker-runc hangs sometimes and needs to be killed in order for the rest of the system to continue to work properly.  Because we run Kubernetes, the most noticeable symptom for us is kubelet on the host would start to show PLEG timeout.  It appears the docker-shim responsible for the runc and the container stops responding.

We can still interact with docker for the most part, and I don't believe we see issues other than kubelet not able to report container events.  `docker ps -a` would show the container ID with status created (but not `docker ps`, but `docker inspect` the container would hang.  `docker-containerd-ctr` shows the container in stopped state.

Another thing that is strange is so far the affected container is kubernetes' infra container.  I have yet to see other container being left in such state.

**Steps to reproduce the issue:**

Can't reproduce it reliably, but happens several times a day across 150 or so docker installations.

**Describe the results you received:**

docker-runc does not terminate, and `docker inspect <cid>` hangs.

**Describe the results you expected:**

docker-runc should terminate, and `docker inspect <cid>` should not hang.

**Additional information you deem important (e.g. issue happens only occasionally):**

Issue happens occasionally, but killing the hung docker-runc process restores the system.

Here is what strace of the stuck docker-runcc looks like

```
# strace -p 43620
strace: Process 43620 attached
openat(AT_FDCWD, "/var/run/docker/runtime-runc/moby/fc2da052c68f4f0a120184ad8eea49fec3e903dd6da9db848bb722f94fb25ba4/exec.fifo", O_RDONLY|O_CLOEXEC
```

**Output of `docker version`:**

```
Client:
 Version:	17.12.0-ce
 API version:	1.35
 Go version:	go1.9.2
 Git commit:	c97c6d6
 Built:	Wed Dec 27 20:10:14 2017
 OS/Arch:	linux/amd64

Server:
 Engine:
  Version:	17.12.0-ce
  API version:	1.35 (minimum version 1.12)
  Go version:	go1.9.2
  Git commit:	c97c6d6
  Built:	Wed Dec 27 20:12:46 2017
  OS/Arch:	linux/amd64
  Experimental:	false
```

**Output of `docker info`:**

```
Containers: 15
 Running: 13
 Paused: 0
 Stopped: 2
Images: 13
Server Version: 17.12.0-ce
Storage Driver: devicemapper
 Pool Name: docker-images
 Pool Blocksize: 65.54kB
 Base Device Size: 10.74GB
 Backing Filesystem: xfs
 Udev Sync Supported: true
 Data Space Used: 3.437GB
 Data Space Total: 797.8GB
 Data Space Available: 794.4GB
 Metadata Space Used: 8.884MB
 Metadata Space Total: 566.2MB
 Metadata Space Available: 557.3MB
 Thin Pool Minimum Free Space: 79.78GB
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.140-RHEL7 (2017-05-03)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 89623f28b87a6004d4b785663257362d1658a729
runc version: b2567b37d7b75eb4cf325b77297b140ea686ce8f
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
 selinux
Kernel Version: 4.14.13-1.el7.elrepo.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 56
Total Memory: 125.8GiB
Name: bnbvxk2
ID: L3JD:VTHD:IQKY:E5IJ:TA3F:PBKH:XYYQ:APCB:LWL3:SHVZ:QKMU:ACJ3
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
 techops.site=slukd1
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: true
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Servers are all bare-metal running CentOS 7.4.1708 with kernel 4.14.11 to 4.14.13.

**Comment by thaJeztah** [comment]
ping @crosbymichael @stevvooe

**Comment by phsiao** [comment]
We can reproduce this problem consistently using a pod creation stress testing job.

```
apiVersion: batch/v1
kind: Job
metadata:
  name: uname-stress-job
spec:
  template:
    spec:
      nodeSelector:
        kubernetes.io/hostname: test-node
      containers:
      - name: uname
        image: alpine
        command: ["uname", "-r"]
      restartPolicy: Never
  backoffLimit: 1
  completions: 1000
  parallelism: 10
```

On a baremetal server this typically finishes in 25-30miin, with docker-runc hangs 3-4 times till our monitoring process comes in and kill those hung docker-runc process.   Without intervention pods on that server would be stuck in ContainerCreating mode for a long time (never waited long enough for it to self-recover, if ever).

Compared to 17.09, where the same stress test results in symptoms like #33820 and requires our monitoring process to come in and restart dockerd, it is a slightly better misbehavior.

**Comment by cpuguy83** [comment]
> docker inspect the container would hang

This is interesting.
Can you get a goroutine stack dump of dockerd?
This is possible by hitting /debug/pprof/goroutine?debug=2 on the docker daemon (debug mode not required).

**Comment by phsiao** [comment]
Upload as a gist at https://gist.github.com/phsiao/b3b371436ac6f254842ecceca56932b7

On the process side

```
# ps ax | grep 34f4152a3721
 6514 ?        Sl     0:00 docker-containerd-shim -namespace moby -workdir /var/lib/docker/containerd/daemon/io.containerd.runtime.v1.linux/moby/34f4152a372146b0c001f47f2b71fb21c7cdd8f7690772fc3b31f8a3219d45b7 -address /var/run/docker/containerd/docker-containerd.sock -containerd-binary /usr/bin/docker-containerd -runtime-root /var/run/docker/runtime-runc
 6629 ?        Sl     0:00 docker-runc --root /var/run/docker/runtime-runc/moby --log /run/docker/containerd/daemon/io.containerd.runtime.v1.linux/moby/34f4152a372146b0c001f47f2b71fb21c7cdd8f7690772fc3b31f8a3219d45b7/log.json --log-format json start 34f4152a372146b0c001f47f2b71fb21c7cdd8f7690772fc3b31f8a3219d45b7
```

and 

```
# strace -p 6629
strace: Process 6629 attached
openat(AT_FDCWD, "/var/run/docker/runtime-runc/moby/34f4152a372146b0c001f47f2b71fb21c7cdd8f7690772fc3b31f8a3219d45b7/exec.fifo", O_RDONLY|O_CLOEXEC
```

`kill 6629` would recover the system into working state again.

**Comment by cpuguy83** [comment]
@phsiao Can you also grab the stack trace from containerd?

```
curl --unix-socket /var/run/docker/containerd/docker-containerd-debug.sock http://./debug/pprof/goroutine?debug=2
```

**Comment by cpuguy83** [comment]
Or alternatively, if you don't have a new-ish version of curl...
```
docker-containerd-ctr pprof --debug-socket /var/run/docker/containerd/docker-containerd-debug.sock goroutines
```

**Comment by crosbymichael** [comment]
tracking a fix in runc here:

https://github.com/opencontainers/runc/pull/1698

**Comment by phsiao** [comment]
Wondering if there is a plan to backport opencontainers/runc#1698 to 17.12 and make a release?

**Comment by phsiao** [comment]
Testing the patch in opencontainers/runc#1698 by replacing the patched file on top of master, and got this in docker info

```
runc version: ab4a8191679806893207d9c1e507a77ff46dafaa-dirty (expected: b2567b37d7b75eb4cf325b77297b140ea686ce8f)
```

And I have run the ```uname-stress-job``` twice and no intervention is needed for it to complete.   Will continue to test and see if there are still issue.

**Comment by cpuguy83** [comment]
Thanks for testing, it is immensely helpful!

**Comment by tvvignesh** [comment]
Hi. I am facing the error below where it says docker-runc did not terminate successfully.

![docker-err](https://user-images.githubusercontent.com/1165845/42729525-73982cd0-87f8-11e8-8f6a-15d4923292c7.png)

I browsed a bit and tried restarting docker service, still did not work.

Then, I got the PID of the containers by using `docker top` and tried `sudo kill -9 <PID>` and I got permission denied.

I tried doing `docker-compose up` again and tried stopping and still, it did not work.

May I know the reason for this and how to fix it?
